### PR TITLE
Add tests for SubArrays in test/linalg/bunchkaufman.jl

### LINKED ISSUE #14732: Linear algebra test coverage for SubArray
Below is the list of linear algebra test files that we'll need to extend with `SubArray` tests. Please update the list if `SubArray`s are already covered (systematically) for the file
- [ ] `arnoldi.jl`
- [x] `bidiag.jl`
- [x] `bunchkaufman.jl`
- [x] `cholesky.jl`
- [ ] `dense.jl`
- [x] `diagonal.jl`
- [x] `eigen.jl`
- [x] `generic.jl` (see #15145)
- [x] `givens.jl`
- [ ] `lapack.jl`
- [x] `lu.jl`
- [x] `lq.jl`
- [x] `matmul.jl`
- [ ] `pinv.jl`
- [x] `qr.jl`
- [x] `schur.jl`
- [ ] `special.jl`
- [x] `svd.jl`
- [ ] `symmetric.jl`
- [ ] `triangular.jl`
- [x] `tridiag.jl`
- [x] `uniformscaling.jl` (see #14731)

### PR DESCRIPTION
**Pull request by pkofod**
- Added tests for `linalg/bunchkaufman.jl`. Found tiny dispatch bug, as `factorize` didn't dispatch properly for `SubArray`s (changed `Matrix` -> `StridedMatrix`).
- Moved two matrix calculations in `svd.jl`.
- Added tests for `givens.jl`. Again, at loop over `Array` and `SubArray`, here for the matrix `A`, and vector `x`. `SubArray`s seem to pass just fine.

### CODE REVIEW

**Comment by andreasnoack** [comment]
It might be better to make individual pull requests for each of the files or maybe a couple and use #14732 to track the progress. It's easier to review smaller PRs and since most of the changes probably won't depend on each other it shouldn't cause issues to split them up.

**Comment by pkofod** [comment]
I agree, edited to reflect your comment. It should be good to go, let's see if Travis agrees.

**Comment by pkofod** [comment]
Doing the `bunchkaufman.jl` tests, I realized I had placed the `asym` and `apd` matrices in such a way, that they were not `SubArray`s, which made coverage worse than I thought. It didn't matter when I moved them in `svd.jl` (most recent commit), but it was what made me catch the `factorize` bug in `bunchkaufman.jl`.

**Comment by pkofod** [comment]
Added tests for `givens.jl`. Again, at loop over `Array` and `SubArray`, here for the matrix `A`, and vector `x`. `SubArray`s seem to pass just fine.

File: test/linalg/bunchkaufman.jl [hunk 1]
```diff
@@ -23,41 +23,62 @@ bimg  = randn(n,2)/2
...
         end
+        ε = εa = eps(abs(float(one(eltya))))
+
+        for eltyb in (Float32, Float64, Complex64, Complex128, Int)
+            b = eltyb == Int ? rand(1:5, n, 2) : convert(Matrix{eltyb}, eltyb <: Complex ? complex(breal, bimg) : breal)
+            for btype in ("Array", "SubArray")
+                if btype == "Array"
+                    b = b
+                else
+                    b = sub(a, 1:n, 1:2)
```
Comment (tkelman) [inline test/linalg/bunchkaufman.jl]: should this be `sub(b` ?

Comment (pkofod) [inline test/linalg/bunchkaufman.jl]: Of course! Will fix

Comment (pkofod) [inline test/linalg/bunchkaufman.jl]: Fixing it, such that `a-> b` makes
```julia
@test_approx_eq_eps asym * (bc1\b) b 1000ε
```
fail with
```julia
ERROR: LoadError: MethodError: no method matching reinterpret(::Type{Float32}, ::SubArray{Complex{Float32},2,Array{Complex{Float32},2},Tuple{UnitRange{Int64},UnitRange{Int64}},false}, ::Tuple{Int64,Int64})
```
Will try to fix it.

**Comment by pkofod** [comment]
Adding the SubArray tests caused https://github.com/JuliaLang/julia/pull/15354/files#diff-be0f75db178315152e44756ee149ffd8R58 to fail for `eltypeb==Complex64` and `typeb=="SubArray"` together. I fixed this at the `\` level, by calling `parent` on `B`. Thoughts on this?

**Comment by andreasnoack** [comment]
Using `parent` seems correct. LGTM. @tkelman do you have further comments?

**Comment by tkelman** [comment]
Just one - given #15367, should `StridedMatrix` continue to be used rather than `AbstractMatrix` ?

**Comment by pkofod** [comment]
I can change it if it's appropriate, and if there's no reason to restrict it to StridedMatrix, I assume that it is.

Is there an easy way to go back to the first commit and change it? I guess I would branch, reset to the commit, make the change, and then merge with my original branch. Is that the way to go?

**Comment by KristofferC** [comment]
You can do `git rebase -i HEAD~3` where the 3 is the number of commits you want to go back and then pick `edit` for the first commit and do your changes.

**Comment by pkofod** [comment]
Perfect, thank you!

**Comment by andreasnoack** [comment]
I'm not sure if these work well beyond `StridedMatrix`. Some of the methods would probably fail for some `DArray`s or `SparseMatrixCSC`s. That might just suggest that some other methods should be fixed to work correctly for those matrix types, but I think we should merge this as it is and wait for a more specific need for `AbstractMatrix`.

**Comment by timholy** [comment]
The use of `parent` is problematic because the parent may not be the same size as the subarray:
```jl
julia> A = rand(6,5); A = A'*A
5x5 Array{Float64,2}:
 2.79981  1.44259  1.9957    2.33348  1.43793 
 1.44259  1.50449  1.16219   1.20452  1.22158 
 1.9957   1.16219  1.93965   1.61173  0.964557
 2.33348  1.20452  1.61173   1.9936   1.11126 
 1.43793  1.22158  0.964557  1.11126  1.66233 

julia> F = cholfact(A);

julia> v6 = rand(Complex128, 6)
6-element Array{Complex{Float64},1}:
 0.485376+0.020852im 
 0.855035+0.32091im  
 0.152016+0.935278im 
  0.78076+0.0182181im
 0.954256+0.378917im 
 0.982591+0.995221im 

julia> v5 = sub(v6, 1:5)
5-element SubArray{Complex{Float64},1,Array{Complex{Float64},1},Tuple{UnitRange{Int64}},true}:
 0.485376+0.020852im 
 0.855035+0.32091im  
 0.152016+0.935278im 
  0.78076+0.0182181im
 0.954256+0.378917im 

julia> F\v5
ERROR: DimensionMismatch("new dimensions (2,5) must be consistent with array size 12")
 in reinterpret(::Type{Float64}, ::Array{Complex{Float64},1}, ::Tuple{Int64,Int64}) at ./array.jl:94
 in \(::Base.LinAlg.Cholesky{Float64,Array{Float64,2}}, ::SubArray{Complex{Float64},1,Array{Complex{Float64},1},Tuple{UnitRange{Int64}},true}) at ./linalg/factorization.jl:27
 in eval(::Module, ::Any) at ./boot.jl:243
```

**Comment by andreasnoack** [comment]
Maybe the original version was correct if we can define `reinterepret` for `SubArray`.

**Comment by pkofod** [comment]
Well that is obviously not intended. To to be sure, is this something you're working on @timholy or were you just pointing it out?

**Comment by timholy** [comment]
Haven't started trying to fix it yet. It was going to get in my way (so I might have fixed it quickly), but then I had to take a step back and think some more.

**Comment by timholy** [comment]
Oh, and @andreasnoack: since `reinterpret` can reshape an array (changing the dimensionality), and `reshape` doesn't currently compose with `SubArray`, I suspect this would be a hard path. (But maybe someday.)

**Comment by pkofod** [comment]
Okay, if I can help let me know.

**Comment by timholy** [comment]
Oh, I meant "think some more" about my BIG SCARY problem, not this specific issue. Sorry for the confusion :smile:.

**Comment by pkofod** [comment]
Oh, okay! So the bug is still there? I'm confused, sorry :)

**Comment by timholy** [comment]
Yes, AFAIK the bug is still there, and I'm not working on it.

**Pull request merged**